src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -0,0 +1,488 @@
+        private static bool IsLowWordSurrogate(uint @char)
+            => (@char & 0xF800U) == 0xD800U;
+
+        public static SequenceValidity PeekFirstSequence(ReadOnlySpan<byte> data, out int numBytesConsumed, out UnicodeScalar scalarValue)
+        {
+            // This method is implemented to match the behavior of System.Text.Encoding.UTF8 in terms of
+            // how many bytes it consumes when reporting invalid sequences. The behavior is as follows:
+            //
+            // - Some bytes are *always* invalid (ranges [ C0..C1 ] and [ F5..FF ]), and when these
+            //   are encountered it's an invalid sequence of length 1.
+            //
+            // - Multi-byte sequences which are overlong are reported as an invalid sequence of length 2,
+            //   since per the Unicode Standard Table 3-7 it's always possible to tell these by the second byte.
+            //   Exception: Sequences which begin with [ C0..C1 ] are covered by the above case, thus length 1.
+            //
+            // - Multi-byte sequences which are improperly terminated (no continuation byte when one is
+            //   expected) are reported as invalid sequences up to and including the last seen continuation byte.
+
+            scalarValue = UnicodeScalar.s_replacementChar;
+
+            if (data.IsEmpty)
+            {
+                // No data to peek at
+                numBytesConsumed = 0;
+                return SequenceValidity.Empty;
+            }
+
+            byte firstByte = data[0];
+
+            if (IsAsciiValue(firstByte))
+            {
+                // ASCII byte = well-formed one-byte sequence.
+                scalarValue = UnicodeScalar.CreateWithoutValidation(firstByte);
+                numBytesConsumed = 1;
+                return SequenceValidity.WellFormed;
+            }
+
+            if (!IsInRangeInclusive(firstByte, (byte)0xC2U, (byte)0xF4U))
+            {
+                // Standalone continuation byte or "always invalid" byte = ill-formed one-byte sequence.
+                goto InvalidOneByteSequence;
+            }
+
+            // At this point, we know we're working with a multi-byte sequence,
+            // and we know that at least the first byte is potentially valid.
+
+            if (data.Length < 2)
+            {
+                // One byte of an incomplete multi-byte sequence.
+                goto OneByteOfIncompleteMultiByteSequence;
+            }
+
+            byte secondByte = data[1];
+
+            if (!IsUtf8ContinuationByte(secondByte))
+            {
+                // One byte of an improperly terminated multi-byte sequence.
+                goto InvalidOneByteSequence;
+            }
+
+            if (firstByte < (byte)0xE0U)
+            {
+                // Well-formed two-byte sequence.
+                scalarValue = UnicodeScalar.CreateWithoutValidation((((uint)firstByte & 0x1FU) << 6) | ((uint)secondByte & 0x3FU));
```

Comment:
What is `UnicodeScalar`, and why can't we use `Rune`?  Isn't it intended for situations like this?

Reply:
cc: @GrabYourPitchforks